sdk/core/azure-core/inc/azure/core/internal/log.hpp

```diff
@@ -3,9 +3,49 @@
 
 #pragma once
 
-#include "azure/core/logging/logging.hpp"
+#include "azure/core/dll_import_export.hpp"
+#include "azure/core/logger.hpp"
 
-namespace Azure { namespace Core { namespace Logging { namespace Internal {
-  bool ShouldLog(LogLevel level);
-  void Log(LogLevel level, std::string const& message);
-}}}} // namespace Azure::Core::Logging::Internal
+#include <atomic>
+#include <type_traits>
+
+namespace Azure { namespace Core { namespace Internal {
+  class Log {
+    using LogLevelInt = std::underlying_type<Logger::Level>::type;
+
+    static_assert(
+        std::is_same<int, LogLevelInt>::value == true && ATOMIC_INT_LOCK_FREE == 2,
+        "Logger::Level values must be representable as lock-free");
+
+    static_assert(ATOMIC_BOOL_LOCK_FREE == 2, "atomic<bool> must be lock-free");
+
+    static AZ_CORE_DLLEXPORT std::atomic<bool> g_isLoggingEnabled;
+    static AZ_CORE_DLLEXPORT std::atomic<LogLevelInt> g_logLevel;
+
+    static Logger::Listener GetLogListener();
+
+    static Logger::Listener GetLogListener(Logger::Level level)
+    {
+      return g_isLoggingEnabled && static_cast<LogLevelInt>(level) <= g_logLevel
+          ? GetLogListener()
+          : Logger::Listener(nullptr);
+    }
+
+    Log() = delete;
+    ~Log() = delete;
+
+  public:
+    static bool ShouldWrite(Logger::Level level) { return GetLogListener(level) != nullptr; }
```

Comment:
This should not be public - customers can't call it

Reply:
This is in the class that is in `Internal` namespace, customers should not be calling it.

Reply:
I don't understand. How do customers call SetLogLevel which is also in this Internal class?